server/src/main/java/org/elasticsearch/cluster/SnapshotsInProgress.java

```diff
@@ -147,9 +153,10 @@ public Entry snapshot(final Snapshot snapshot) {
      */
     public Map<RepositoryShardId, Set<ShardGeneration>> obsoleteGenerations(String repository, SnapshotsInProgress old) {
         final Map<RepositoryShardId, Set<ShardGeneration>> obsoleteGenerations = new HashMap<>();
+        final List<Entry> updatedSnapshots = forRepo(repository);
```

Comment:
This is the trick pretty much, the lookup from the map is almost all of the runtime here. Iterating an immutable list is comparatively irrelevant it seems (and this is at 5k concurrent snapshots):

![Uploading image.png…]()